Mixed Java & Scala 3.x compilation broke with release 4.7.2 and newer

It looks like mixed compilation with Java 8 and Scala 3.x  (Scala 3.2.2 in my case) is broken. 

The compilation fails with:

```
[INFO] compiling 2 Scala sources and 54 Java sources to /home/roman/dev/include-cloud/datacenter/target/classes ...
[ERROR] : 1.8 is not a valid choice for -java-output-version
[INFO] :   scalac -help  gives more information
[ERROR] one error found
```

The issue seems to be introduced by this change: https://github.com/davidB/scala-maven-plugin/commit/ede3d72e206b8d9a7ad07d28fc27863ccc7264d6

The rough issue: `-release 1.8` is added to the Scalac command, which it doesn't like. The `ScalaMojoSupport.targetOption` does translate the Maven `1.8` Java version to `8`, but only for the Scala 2.x compiler series. 

So, it seems that either the `-release` flag isn't correct for Scala 3.x, or the version translation should be applied to Scacal 3.x as well.

--- a/src/main/java/scala_maven/ScalaMojoSupport.java
+++ b/src/main/java/scala_maven/ScalaMojoSupport.java
@@ -588,32 +588,61 @@ public abstract class ScalaMojoSupport extends AbstractMojo {
     return FileUtils.toMultiPath(classpath);
   }
 
-  static String targetOption(String target, VersionNumber scalaVersion) {
-    if (scalaVersion.major == 2) {
-      if (scalaVersion.minor <= 12) {
-        if (target.equals("1.5") || target.equals("5")) {
-          return "jvm-1.5";
-        } else if (target.equals("1.6") || target.equals("6")) {
-          return "jvm-1.6";
-        } else if (target.equals("1.7") || target.equals("7")) {
-          return "jvm-1.7";
-        } else if (target.equals("1.8") || target.equals("8")) {
-          return "jvm-1.8";
-        } else {
-          // invalid or unsupported option, just ignore
-          return null;
-        }
-      } else if (target.equals("1.5")) {
-        return "5";
-      } else if (target.equals("1.6")) {
-        return "6";
-      } else if (target.equals("1.7")) {
-        return "7";
-      } else if (target.equals("1.8")) {
-        return "8";
-      }
+  private static String computeTargetOption(String target) {
+    if (target.equals("1.5") || target.equals("5")) {
+      return "jvm-1.5";
+    } else if (target.equals("1.6") || target.equals("6")) {
+      return "jvm-1.6";
+    } else if (target.equals("1.7") || target.equals("7")) {
+      return "jvm-1.7";
+    } else if (target.equals("1.8") || target.equals("8")) {
+      return "jvm-1.8";
+    }
+    return null;
+  }
+
+  private static String computeReleaseOptionFromTarget(String target) {
+    if (target.equals("1.5")) {
+      return "5";
+    } else if (target.equals("1.6")) {
+      return "6";
+    } else if (target.equals("1.7")) {
+      return "7";
+    } else if (target.equals("1.8")) {
+      return "8";
     }
     return target;
+  }
+
+  // visible for tests
+  static List<String> computeBytecodeVersionOptions(
+      String target, String release, VersionNumber scalaVersion) {
+    List<String> options = new ArrayList<>();
+    boolean targetIsDefined = StringUtils.isNotEmpty(target);
+    boolean releaseIsDefined = StringUtils.isNotEmpty(release);
+    boolean releaseIsSupported = scalaVersion.compareTo(new VersionNumber("2.12.0")) >= 0;
+    String releaseOrJavaOutputVersionOptionName =
+        scalaVersion.compareTo(new VersionNumber("3.1.2")) >= 0
+            ? "-java-output-version"
+            : "-release";
+
+    if (releaseIsDefined && releaseIsSupported) {
+      // release's default is "maven.compiler.release"'s default, which is null
+      options.add(releaseOrJavaOutputVersionOptionName);
+      options.add(release);
+    } else if (targetIsDefined) {
+      // target's default is "maven.compiler.target"'s default, which is 1.8
+      if (releaseIsSupported) {
+        options.add(releaseOrJavaOutputVersionOptionName);
+        options.add(computeReleaseOptionFromTarget(target));
+      } else {
+        String correctTarget = computeTargetOption(target);
+        if (correctTarget != null) {
+          options.add("-target:" + correctTarget);
+        }
+      }
+    }
+    return options;
   }
 
   protected List<String> getScalacOptions() throws Exception {
@@ -623,35 +652,7 @@ public abstract class ScalaMojoSupport extends AbstractMojo {
       Collections.addAll(options, StringUtils.split(addScalacArgs, "|"));
     }
     options.addAll(getCompilerPluginOptions());
-
-    VersionNumber scalaVersion = findScalaVersion();
-
-    boolean targetIsDefined = StringUtils.isNotEmpty(target);
-    boolean releaseIsDefined = StringUtils.isNotEmpty(release);
-    boolean targetIsDeprecated = scalaVersion.compareTo(new VersionNumber("2.13.9")) >= 0;
-    boolean releaseIsSupported = scalaVersion.compareTo(new VersionNumber("2.12.0")) >= 0;
-
-    // target's default is "maven.compiler.target"'s default, which is 1.8
-    if (targetIsDefined) {
-      String targetOption = targetOption(target, scalaVersion);
-      if (targetOption != null) {
-        if (!targetIsDeprecated) {
-          options.add("-target:" + targetOption);
-        } else if (!releaseIsDefined) {
-          // -target is deprecated in favor of -release
-          // no user-defined release specified
-          // set release instead, so we don't get a deprecation warning
-          options.add("-release");
-          options.add(targetOption);
-        }
-      }
-    }
-
-    // release's default is "maven.compiler.release"'s default, which is null
-    if (releaseIsDefined && releaseIsSupported) {
-      options.add("-release");
-      options.add(release);
-    }
+    options.addAll(computeBytecodeVersionOptions(target, release, findScalaVersion()));
 
     return options;
   }

--- a/src/test/java/scala_maven/ScalaMojoSupportTest.java
+++ b/src/test/java/scala_maven/ScalaMojoSupportTest.java
@@ -4,8 +4,8 @@
  */
 package scala_maven;
 
-import static org.junit.Assert.assertEquals;
-import static org.junit.Assert.assertNull;
+import static java.util.Arrays.asList;
+import static org.junit.Assert.*;
 
 import org.junit.Test;
 
@@ -13,51 +13,103 @@ public class ScalaMojoSupportTest {
 
   @Test
   public void scala2_11_should_generate_prefixed_target() {
-    assertEquals("jvm-1.5", ScalaMojoSupport.targetOption("1.5", new VersionNumber("2.11.12")));
-    assertEquals("jvm-1.5", ScalaMojoSupport.targetOption("5", new VersionNumber("2.11.12")));
-    assertEquals("jvm-1.6", ScalaMojoSupport.targetOption("1.6", new VersionNumber("2.11.12")));
-    assertEquals("jvm-1.6", ScalaMojoSupport.targetOption("6", new VersionNumber("2.11.12")));
-    assertEquals("jvm-1.7", ScalaMojoSupport.targetOption("1.7", new VersionNumber("2.11.12")));
-    assertEquals("jvm-1.7", ScalaMojoSupport.targetOption("7", new VersionNumber("2.11.12")));
-    assertEquals("jvm-1.8", ScalaMojoSupport.targetOption("1.8", new VersionNumber("2.11.12")));
-    assertEquals("jvm-1.8", ScalaMojoSupport.targetOption("8", new VersionNumber("2.11.12")));
+    assertEquals(
+        asList("-target:jvm-1.5"),
+        ScalaMojoSupport.computeBytecodeVersionOptions("1.5", null, new VersionNumber("2.11.12")));
+    assertEquals(
+        asList("-target:jvm-1.5"),
+        ScalaMojoSupport.computeBytecodeVersionOptions("5", null, new VersionNumber("2.11.12")));
+    assertEquals(
+        asList("-target:jvm-1.6"),
+        ScalaMojoSupport.computeBytecodeVersionOptions("1.6", null, new VersionNumber("2.11.12")));
+    assertEquals(
+        asList("-target:jvm-1.6"),
+        ScalaMojoSupport.computeBytecodeVersionOptions("6", null, new VersionNumber("2.11.12")));
+    assertEquals(
+        asList("-target:jvm-1.7"),
+        ScalaMojoSupport.computeBytecodeVersionOptions("1.7", null, new VersionNumber("2.11.12")));
+    assertEquals(
+        asList("-target:jvm-1.7"),
+        ScalaMojoSupport.computeBytecodeVersionOptions("7", null, new VersionNumber("2.11.12")));
+    assertEquals(
+        asList("-target:jvm-1.8"),
+        ScalaMojoSupport.computeBytecodeVersionOptions("1.8", null, new VersionNumber("2.11.12")));
+    assertEquals(
+        asList("-target:jvm-1.8"),
+        ScalaMojoSupport.computeBytecodeVersionOptions("8", null, new VersionNumber("2.11.12")));
   }
 
   @Test
-  public void scala2_11_should_generate_null_for_unsupported_java_versions() {
-    assertNull(ScalaMojoSupport.targetOption("11", new VersionNumber("2.11.12")));
-    assertNull(ScalaMojoSupport.targetOption("17", new VersionNumber("2.11.12")));
+  public void scala2_11_should_generate_nothing_for_unsupported_java_versions() {
+    assertTrue(
+        ScalaMojoSupport.computeBytecodeVersionOptions("11", null, new VersionNumber("2.11.12"))
+            .isEmpty());
+    assertTrue(
+        ScalaMojoSupport.computeBytecodeVersionOptions("17", null, new VersionNumber("2.11.12"))
+            .isEmpty());
   }
 
   @Test
-  public void scala2_12_should_generate_prefixed_target() {
-    assertEquals("jvm-1.5", ScalaMojoSupport.targetOption("1.5", new VersionNumber("2.12.11")));
-    assertEquals("jvm-1.5", ScalaMojoSupport.targetOption("5", new VersionNumber("2.12.11")));
-    assertEquals("jvm-1.6", ScalaMojoSupport.targetOption("1.6", new VersionNumber("2.12.11")));
-    assertEquals("jvm-1.6", ScalaMojoSupport.targetOption("6", new VersionNumber("2.12.11")));
-    assertEquals("jvm-1.7", ScalaMojoSupport.targetOption("1.7", new VersionNumber("2.12.11")));
-    assertEquals("jvm-1.7", ScalaMojoSupport.targetOption("7", new VersionNumber("2.12.11")));
-    assertEquals("jvm-1.8", ScalaMojoSupport.targetOption("1.8", new VersionNumber("2.12.11")));
-    assertEquals("jvm-1.8", ScalaMojoSupport.targetOption("8", new VersionNumber("2.12.11")));
+  public void scala2_12_should_generate_release() {
+    assertEquals(
+        asList("-release", "8"),
+        ScalaMojoSupport.computeBytecodeVersionOptions("1.8", null, new VersionNumber("2.12.11")));
+    assertEquals(
+        asList("-release", "8"),
+        ScalaMojoSupport.computeBytecodeVersionOptions("1.8", "8", new VersionNumber("2.12.11")));
+    assertEquals(
+        asList("-release", "11"),
+        ScalaMojoSupport.computeBytecodeVersionOptions("1.8", "11", new VersionNumber("2.12.11")));
+    assertEquals(
+        asList("-release", "17"),
+        ScalaMojoSupport.computeBytecodeVersionOptions("1.8", "17", new VersionNumber("2.12.11")));
   }
 
   @Test
-  public void scala2_12_should_generate_null_for_unsupported_java_versions() {
-    assertNull(ScalaMojoSupport.targetOption("11", new VersionNumber("2.12.11")));
-    assertNull(ScalaMojoSupport.targetOption("17", new VersionNumber("2.12.11")));
+  public void scala2_13_should_generate_release() {
+    assertEquals(
+        asList("-release", "8"),
+        ScalaMojoSupport.computeBytecodeVersionOptions("1.8", null, new VersionNumber("2.13.10")));
+    assertEquals(
+        asList("-release", "8"),
+        ScalaMojoSupport.computeBytecodeVersionOptions("1.8", "8", new VersionNumber("2.13.10")));
+    assertEquals(
+        asList("-release", "11"),
+        ScalaMojoSupport.computeBytecodeVersionOptions("1.8", "11", new VersionNumber("2.13.10")));
+    assertEquals(
+        asList("-release", "17"),
+        ScalaMojoSupport.computeBytecodeVersionOptions("1.8", "17", new VersionNumber("2.13.10")));
   }
 
   @Test
-  public void scala2_13_should_generate_non_prefixed_target() {
-    assertEquals("5", ScalaMojoSupport.targetOption("1.5", new VersionNumber("2.13.8")));
-    assertEquals("5", ScalaMojoSupport.targetOption("5", new VersionNumber("2.13.8")));
-    assertEquals("6", ScalaMojoSupport.targetOption("1.6", new VersionNumber("2.13.8")));
-    assertEquals("6", ScalaMojoSupport.targetOption("6", new VersionNumber("2.13.8")));
-    assertEquals("7", ScalaMojoSupport.targetOption("1.7", new VersionNumber("2.13.8")));
-    assertEquals("7", ScalaMojoSupport.targetOption("7", new VersionNumber("2.13.8")));
-    assertEquals("8", ScalaMojoSupport.targetOption("1.8", new VersionNumber("2.13.8")));
-    assertEquals("8", ScalaMojoSupport.targetOption("8", new VersionNumber("2.13.8")));
-    assertEquals("11", ScalaMojoSupport.targetOption("11", new VersionNumber("2.13.8")));
-    assertEquals("17", ScalaMojoSupport.targetOption("17", new VersionNumber("2.13.8")));
+  public void scala3_1_1_should_generate_release() {
+    assertEquals(
+        asList("-release", "8"),
+        ScalaMojoSupport.computeBytecodeVersionOptions("1.8", null, new VersionNumber("3.1.1")));
+    assertEquals(
+        asList("-release", "8"),
+        ScalaMojoSupport.computeBytecodeVersionOptions("1.8", "8", new VersionNumber("3.1.1")));
+    assertEquals(
+        asList("-release", "11"),
+        ScalaMojoSupport.computeBytecodeVersionOptions("1.8", "11", new VersionNumber("3.1.1")));
+    assertEquals(
+        asList("-release", "17"),
+        ScalaMojoSupport.computeBytecodeVersionOptions("1.8", "17", new VersionNumber("3.1.1")));
+  }
+
+  @Test
+  public void scala3_1_2_should_generate_java_output_version() {
+    assertEquals(
+        asList("-java-output-version", "8"),
+        ScalaMojoSupport.computeBytecodeVersionOptions("1.8", null, new VersionNumber("3.1.2")));
+    assertEquals(
+        asList("-java-output-version", "8"),
+        ScalaMojoSupport.computeBytecodeVersionOptions("1.8", "8", new VersionNumber("3.1.2")));
+    assertEquals(
+        asList("-java-output-version", "11"),
+        ScalaMojoSupport.computeBytecodeVersionOptions("1.8", "11", new VersionNumber("3.1.2")));
+    assertEquals(
+        asList("-java-output-version", "17"),
+        ScalaMojoSupport.computeBytecodeVersionOptions("1.8", "17", new VersionNumber("3.1.2")));
   }
 }

